src/Mocking.jl

```diff
@@ -34,125 +37,57 @@ function enable(; force::Bool=false)
-# in as an array into Patch. At runtime the types and function names will be fully
-# qualified
-
-# We can support optional parameters and keywords by using generic functions on
-# 0.4
-
-function convert(::Type{Expr}, p::Patch)
-    exprs = Expr[]
-
-    # Generate imports for all required modules
-    for m in p.modules
-        bindings = splitbinding(m)
-
-        :Main in bindings && error("Mocking cannot handle bindings from Main.")
-
-        for i in 1:length(bindings)
-            import_expr = Expr(:import, Expr(:., bindings[1:i]...))
-            push!(exprs, import_expr)
-        end
-    end
-
-    # Generate the new method which will call the user's patch function. We need to perform
-    # this call instead of injecting the body expression to support closures.
-    sig, body = p.signature, p.body
-    params = call_parameters(sig)
-    push!(exprs, Expr(:(=), sig, Expr(:block, Expr(:call, body, params...))))
-
-    return Expr(:block, exprs...)
+    target::Callable
+    alternate::Function
 end
 
 macro patch(expr::Expr)
-    if expr.head == :function
-        name = expr.args[1].args[1]
-        params = expr.args[1].args[2:end]
-        body = expr.args[2]
-
-    # Short-form function syntax
-    elseif expr.head == :(=) && expr.args[1].head == :call
-        name = expr.args[1].args[1]
+    # Expect a named function in long-form or short-form
+    if expr.head === :function || (expr.head === :(=) && expr.args[1].head == :call)
+        target_name = expr.args[1].args[1]
         params = expr.args[1].args[2:end]
         body = expr.args[2]
-
-    # Anonymous function syntax
-    # elseif expr.head == :(->)
-        # TODO: Determine how this could be supported
+    elseif expr.head == :(->)
+        throw(ArgumentError("expression needs to be a named function"))
     else
         throw(ArgumentError("expression is not a function definition"))
     end
 
-    signature = Expr(:call, name, params...)
-
-    # Determine the bindings used in the signature
-    bindings = Bindings(signature)
+    patch_name = target_name isa Expr ? target_name.args[2].value : target_name
 
     # Need to evaluate the body of the function in the context of the `@patch` macro in
     # order to support closures.
     # func = Expr(:(->), Expr(:tuple, params...), body)
-    func = Expr(:(=), Expr(:call, gensym(), params...), body)
-
-    # Generate a translation between the external bindings and the runtime types and
-    # functions. The translation will be used to revise all bindings to be absolute.
-    translations = [Expr(:call, :(=>), QuoteNode(b), b) for b in bindings.external]
+    func = Expr(:(=), Expr(:call, gensym(patch_name), params...), body)
 
-    return esc(:(Mocking.Patch( $(QuoteNode(signature)), $func, Dict($(translations...)) )))
+    return esc(:(Mocking.Patch($target_name, $func)))
 end
 
 struct PatchEnv
-    mod::Module
+    mapping::Dict{Callable,Vector{Function}}
```

Comment:
Valid use case for a `LittleDict`,  since it will be small

Reply:
I totally agree. I'd like to do some benchmarking first before making the switch.

Reply:
LittleDIct is most happpy if you give it 1 Tuple of Keys and 1 Tuple of values.
So it doesn't have to construct anything and is fully immutable)

Reply:
I looked into using `LittleDict` and benchmarked it via:
```julia
using Mocking, BenchmarkTools
funcs = map(i -> @eval($(Symbol("f$i"))(x, y=1, z=2) = $i), 1:5)
pe = Mocking.PatchEnv(Patch(f, f) for f in funcs)
@btime Mocking.get_alternate($pe, f3)
@btime Mocking.get_alternate($pe, +)
```

Using `Dict{Callable,Vector{Function}}`
```julia
julia> @btime Mocking.get_alternate($pe, f3)
  1.948 μs (10 allocations: 528 bytes)

julia> @btime Mocking.get_alternate($pe, +)
  171.159 ns (0 allocations: 0 bytes)
```

Using `LittleDict{Callable,Vector{Function}}`
```julia
julia> @btime Mocking.get_alternate($pe, f3)
  3.332 μs (10 allocations: 528 bytes)

julia> @btime Mocking.get_alternate($pe, +)
  775.616 ns (0 allocations: 0 bytes)
```

Using `LittleDict{Tuple{Callable},Tuple{Vector{Function}}}`
```julia
julia> @btime Mocking.get_alternate($pe, f3)
  2.517 μs (10 allocations: 528 bytes)

julia> @btime Mocking.get_alternate($pe, +)
  320.142 ns (0 allocations: 0 bytes)
```

Testing `LittleDict{Tuple{Callable},Tuple{Vararg{Function}}}` is a bit tricky to implement and based upon the results I'm seeing I think `Dict` is better here.